tensilelite/Tensile/KernelWriterAssembly.py

```diff
@@ -7585,9 +7627,11 @@ def globalReadGuardKBody(tP, optParams = None):
                       eccOffset = _getEccOffset(tP["globalReadInstruction"].totalWidth, bpr=self.states.bpr, bpe=eccBpe, \
                         glvw=tP["glvw"], idx=loopCnt, numVgprG2L=numVgprG2L)
                 elif dataType.isHalf() or dataType.isBFloat16():
-                  if tP["glvw"]>1 and kernel["AssertSummationElementMultiple"] % 2 == 0:
+                  if tP["glvw"]>1 and kernel["AssertSummationElementMultiple"] % 2 == 0 and not isLds:
                   # Pack two FP16 values into a single load dword x2
                     numElementsPerLoad = 2
+                  elif isLds and kernel["AssertSummationElementMultiple"] % 2 == 0:
+                    numElementsPerLoad = kernel["GlobalReadVectorWidth%c"%tc]
```

Comment:
Can we really do this?
Doesn't this cause out of boundary access?
For example, if ASEM=2, GRVW is 4 and K is not multiple of 4, I think this does out of boundary access since each global load accesses 4 elements at once.

Reply:
Are you sure?
That is different from my understanding.
Let's say, address=0x1000, OOB is 0x1004, GRVW=4 (fp16, 8byte access per thread),
in this case, global read instruction accesses from 0x1000 to 0x1007.
My understanding is buffer load limit (Srd+2) does not limit the OOB access if the limit is within global load unit.

If OOB access limit really works in this case, I think all edge related logics (Edge shift, narrower load for TailLoop) are unnecessary.

Reply:
Let me double check and get back to you. I believe all dword/dwordx2/dwordx4 have read granularity of 32 bits and in the example you provided the vgprs containing elements in 0x1004-0x1007 would be zero.

Reply:
OK. Please dump the register value and make sure if it is 0 or not.
By the way, setting value 0 to OOB does not work with dtl option.

Reply:
Confirmed, size record determines OOB cut-off with 32b granularity (for dword{x2,x4}). 

Comment: the 32b granularity makes directly reading 2 bytes not possible with dword{x2,x4} loads - i.e. if OOB is 0x1002, load will return 0 when trying to read 2 bytes from 0x1000. My previous [PR](https://github.com/ROCm/hipBLASLt/pull/2046) was trying to workaround this. 

> setting value 0 to OOB does not work with dtl option.

What do you mean by this?

Reply:
OK. Thanks for confirmation.
I did not know that.

Regarding this
> setting value 0 to OOB does not work with dtl option.

Without dtl option, global load value in register from OOB memory will be 0.
However, dtl case, looks like 0 is not written to LDS (if I remember correctly).